File: go/pkg/discovery/cache.go
package discovery

import (
	"container/list"
	"context"
	"sync"
	"time"

	"golang.org/x/sync/singleflight"
)

// cacheEntry is a cached configuration with its expiry instant. key duplicates
// the issuer URL the entry is stored under so an LRU eviction can drop it from
// the lookup map by key.
type cacheEntry struct {
	key       string
	cfg       *ProviderConfiguration
	expiresAt time.Time
}

// cache is a bounded, TTL cache for discovery documents keyed by issuer URL.
// Concurrent fetches for the same issuer are deduplicated via singleflight so
// only one HTTP request is made when the cache is empty or expired.
//
// The cache is bounded by a max-entries LRU: without a bound it grows one entry
// per distinct issuer forever, so a caller driven to resolve many distinct
// issuers (a multi-tenant gateway, an attacker-supplied issuer header) leaks
// memory unbounded (a DoS). Recency is tracked in order — a read hit or a store
// moves the entry to the front, and eviction pops the back — so the entry
// evicted under pressure is the least recently *used*, not merely the least
// recently inserted. FIFO would let an attacker reading distinct issuers in
// insertion order evict a legitimately hot entry.
type cache struct {
	mu sync.RWMutex

	// entries maps issuer URL -> its element in order; each element's Value is a
	// *cacheEntry. The single map+list pair gives O(1) lookup, recency promotion
	// and eviction, all serialized by mu.
	entries map[string]*list.Element

	// order tracks access recency: most-recently-used at the front,
	// least-recently-used at the back. Eviction removes from the back.
	order *list.List

	group singleflight.Group

	// now returns the current time; overridable in tests to drive TTL expiry
	// deterministically (DISC-005).
	now func() time.Time
}

// globalCache backs the package-level [FetchConfiguration].
var globalCache = newCache()

// newCache returns an empty cache using the wall clock.
func newCache() *cache {
	return &cache{
		entries: make(map[string]*list.Element),
		order:   list.New(),
		now:     time.Now,
	}
}

// lookup returns the cached configuration for key if present and unexpired,
// promoting the entry to most-recently-used on a hit. Promotion mutates the
// recency list, so this takes the exclusive lock rather than a read lock; the
// critical section is a map probe and an O(1) list move. An expired entry is
// neither promoted nor removed here — it is overwritten by the next store or
// evicted under pressure.
func (c *cache) lookup(key string) (*ProviderConfiguration, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()
	elem, ok := c.entries[key]
	if !ok {
		return nil, false
	}
	entry := elem.Value.(*cacheEntry)
	if !c.now().Before(entry.expiresAt) {
		return nil, false
	}
	c.order.MoveToFront(elem)
	return entry.cfg, true
}

// store records cfg for key with the supplied TTL, marks the entry
// most-recently-used, and enforces the max-entries bound. A maxEntries <= 0
// disables the bound (unbounded) — the backward-compatible escape hatch.
func (c *cache) store(key string, cfg *ProviderConfiguration, ttl time.Duration, maxEntries int) {
	c.mu.Lock()
	defer c.mu.Unlock()
	entry := &cacheEntry{key: key, cfg: cfg, expiresAt: c.now().Add(ttl)}
	if elem, ok := c.entries[key]; ok {
		elem.Value = entry
		c.order.MoveToFront(elem)
	} else {
		c.entries[key] = c.order.PushFront(entry)
	}
	c.evictLocked(maxEntries)
}

// evictLocked drops least-recently-used entries until at most maxEntries remain.
// It must be called with mu held. A maxEntries <= 0 disables eviction.
func (c *cache) evictLocked(maxEntries int) {
	if maxEntries <= 0 {
		return
	}
	for c.order.Len() > maxEntries {
		back := c.order.Back()
		if back == nil {
			return
		}
		evicted := c.order.Remove(back).(*cacheEntry)
		delete(c.entries, evicted.key)
	}
}

// clear drops every cached configuration.
func (c *cache) clear() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.entries = make(map[string]*list.Element)
	c.order = list.New()
}

// fetch returns the cached configuration for issuerURL or fetches it, caching
// the result. Concurrent misses for the same issuer collapse to one fetch.
//
// Concurrency semantics: when several callers miss the cache simultaneously,
// only the first triggers the HTTP request and the rest wait on its result.
// The shared fetch runs on a context detached from any single caller (via
// [context.WithoutCancel]), so one caller cancelling or timing out cannot
// poison the request other callers depend on; each caller still observes its
// own context cancellation through the select below. Because the shared fetch
// is detached, a deadline carried only on a caller's context does not bound the
// underlying request (the caller still unblocks at its deadline, but the fetch
// runs to the configured [WithTimeout] or the default request timeout). The
// timeout, cache TTL and max-entries bound applied to the shared fetch are those
// of the caller that wins the flight (first-wins) — callers needing distinct
// timeouts should not rely on the shared cache for that guarantee.
func (c *cache) fetch(ctx context.Context, issuerURL string, cfg *config) (*ProviderConfiguration, error) {
	// DISC-004: serve a fresh cache entry without any HTTP request.
	if doc, ok := c.lookup(issuerURL); ok {
		return doc, nil
	}

	// Singleflight collapses concurrent misses into one in-flight request.
	// DoChan (not Do) lets each caller honour its own context independently.
	ch := c.group.DoChan(issuerURL, func() (interface{}, error) {
		// Re-check under the flight in case another goroutine just populated
		// the cache (DISC-005 boundary).
		if doc, ok := c.lookup(issuerURL); ok {
			return doc, nil
		}
		// Detach from the winning caller's context so its cancellation does
		// not abort the fetch the other waiters share. fetchAndValidate still
		// bounds the request with the configured (or default) timeout.
		doc, err := fetchAndValidate(context.WithoutCancel(ctx), issuerURL, cfg)
		if err != nil {
			return nil, err
		}
		// Only successful fetches are cached.
		c.store(issuerURL, doc, cfg.cacheTTL, cfg.maxEntries)
		return doc, nil
	})

	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	case res := <-ch:
		if res.Err != nil {
			return nil, res.Err
		}
		return res.Val.(*ProviderConfiguration), nil
	}
}

File: rust/src/discovery/cache.rs
//! Bounded TTL cache for discovery documents, keyed by issuer URL.
//!
//! Backed by a [`tokio::sync::RwLock`] so concurrent readers share a fresh
//! entry without contention (DISC-004) while a refresh takes the write lock
//! (DISC-005).
//!
//! The map is keyed by issuer URL; without a bound it grows one entry per
//! distinct issuer forever. A multi-tenant gateway or attacker-supplied issuer
//! header could accumulate entries until the process exhausts memory
//! (unbounded-growth DoS). A max-entries LRU bound caps the map: insertion
//! evicts the least-recently-*used* entry once the cap is exceeded, and read
//! hits refresh recency so a hot entry is not evicted by a flood of cold
//! issuers. Mirrors the Python reference (`py_identity_model.core.jwks_cache`,
//! default 64). A cap of `0` disables the bound (explicit unbounded escape
//! hatch).

use std::collections::HashMap;
use std::sync::atomic::{AtomicU64, Ordering};
use std::time::{Duration, Instant};

use tokio::sync::RwLock;

use super::metadata::ProviderMetadata;

/// A cached document paired with the instant it expires. `expires_at` is
/// `None` when the TTL is so large that `Instant::now() + ttl` would overflow
/// (e.g. `Duration::MAX`), in which case the entry never expires.
struct CacheEntry {
    metadata: ProviderMetadata,
    expires_at: Option<Instant>,
    /// Monotonic recency tick; the higher the value, the more recently the entry
    /// was used. Bumped on every read hit and on insert. Interior-mutable
    /// ([`AtomicU64`]) so a read hit can refresh recency while holding only a
    /// shared read lock, preserving the concurrent-reader property (DISC-004).
    last_access: AtomicU64,
}

/// A bounded TTL cache mapping an issuer URL to its parsed [`ProviderMetadata`].
pub(crate) struct Cache {
    entries: RwLock<HashMap<String, CacheEntry>>,
    /// Maximum number of distinct issuer entries retained. `0` means unbounded
    /// (no eviction).
    max_entries: usize,
    /// Source of the monotonic recency ticks stored in each entry.
    access_counter: AtomicU64,
}

impl Cache {
    /// Returns an empty cache bounded to at most `max_entries` distinct issuers.
    /// A `max_entries` of `0` disables the bound (unbounded).
    pub(crate) fn new(max_entries: usize) -> Self {
        Self {
            entries: RwLock::new(HashMap::new()),
            max_entries,
            access_counter: AtomicU64::new(0),
        }
    }

    /// Returns the next monotonic recency tick.
    fn next_tick(&self) -> u64 {
        self.access_counter.fetch_add(1, Ordering::Relaxed)
    }

    /// Returns the cached metadata for `key` if present and unexpired
    /// (DISC-004); returns `None` once the TTL has elapsed (DISC-005). A hit
    /// refreshes the entry's LRU recency so it is evicted last.
    pub(crate) async fn get(&self, key: &str) -> Option<ProviderMetadata> {
        let entries = self.entries.read().await;
        let entry = entries.get(key)?;
        // A `None` expiry never elapses; otherwise the entry is fresh until then.
        if entry.expires_at.is_none_or(|at| Instant::now() < at) {
            // Refresh recency under the shared read lock via the atomic tick.
            entry.last_access.store(self.next_tick(), Ordering::Relaxed);
            Some(entry.metadata.clone())
        } else {
            None
        }
    }

    /// Stores `metadata` for `key`, expiring `ttl` from now, then evicts the
    /// least-recently-used entries until the cache is within its bound.
    pub(crate) async fn put(&self, key: String, metadata: ProviderMetadata, ttl: Duration) {
        // `checked_add` guards against an overflow panic for a very large TTL
        // (e.g. `Duration::MAX`); `None` means the entry never expires.
        let expires_at = Instant::now().checked_add(ttl);
        let mut entries = self.entries.write().await;
        // Take the recency tick *after* acquiring the write lock so tick order
        // matches actual insertion order under concurrent puts (ticking before
        // the lock could hand a later insert a smaller tick, skewing the LRU
        // victim). Matches the Go port's `store`.
        let tick = self.next_tick();
        entries.insert(
            key,
            CacheEntry {
                metadata,
                expires_at,
                last_access: AtomicU64::new(tick),
            },
        );
        self.evict_lru(&mut entries);
    }

    /// Evicts least-recently-used entries until `entries` is within
    /// `max_entries`. A `max_entries` of `0` is unbounded and evicts nothing.
    /// Runs under the caller's write lock.
    fn evict_lru(&self, entries: &mut HashMap<String, CacheEntry>) {
        if self.max_entries == 0 {
            return;
        }
        while entries.len() > self.max_entries {
            // The least-recently-used entry has the smallest recency tick. Ties
            // are effectively impossible: every tick is a distinct value handed
            // out by `next_tick`. The scan is O(n) with n <= max_entries.
            let lru_key = entries
                .iter()
                .min_by_key(|(_, entry)| entry.last_access.load(Ordering::Relaxed))
                .map(|(key, _)| key.clone());
            match lru_key {
                Some(key) => {
                    entries.remove(&key);
                }
                None => break,
            }
        }
    }

    /// Test-only: number of cached entries.
    #[cfg(test)]
    pub(crate) async fn entry_count(&self) -> usize {
        self.entries.read().await.len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// A minimal metadata value for cache-mechanics tests; contents are
    /// irrelevant here (LRU keys on the issuer, not the payload). Every required
    /// field defaults to empty, so `{}` deserializes.
    fn metadata() -> ProviderMetadata {
        serde_json::from_str("{}").expect("empty metadata deserializes")
    }

    const TTL: Duration = Duration::from_secs(60);

    // The cache never exceeds its configured cap no matter how many distinct
    // issuers are inserted (unbounded-growth DoS bound).
    #[tokio::test]
    async fn size_never_exceeds_cap() {
        let cache = Cache::new(3);
        for i in 0..25 {
            cache
                .put(format!("https://issuer/{i}"), metadata(), TTL)
                .await;
            assert!(
                cache.entry_count().await <= 3,
                "size exceeded cap after inserting entry {i}"
            );
        }
        assert_eq!(cache.entry_count().await, 3);
    }

    // LRU eviction order: the least-recently-*used* entry is evicted first, and a
    // read hit refreshes recency so a hot entry survives a later cold insert.
    #[tokio::test]
    async fn evicts_least_recently_used_and_read_refreshes_recency() {
        let cache = Cache::new(2);
        cache.put("a".to_string(), metadata(), TTL).await;
        cache.put("b".to_string(), metadata(), TTL).await;

        // Read "a" so it becomes more recently used than "b".
        assert!(cache.get("a").await.is_some());

        // Inserting "c" overflows the cap; the LRU entry is now "b".
        cache.put("c".to_string(), metadata(), TTL).await;

        assert!(cache.get("a").await.is_some(), "recently-read a survives");
        assert!(cache.get("c").await.is_some(), "newest c present");
        assert!(
            cache.get("b").await.is_none(),
            "least-recently-used b evicted"
        );
        assert_eq!(cache.entry_count().await, 2);
    }

    // A cap of 0 disables the bound entirely (backward-compat unbounded mode).
    #[tokio::test]
    async fn zero_cap_is_unbounded() {
        let cache = Cache::new(0);
        for i in 0..100 {
            cache.put(format!("issuer-{i}"), metadata(), TTL).await;
        }
        assert_eq!(cache.entry_count().await, 100, "cap 0 evicts nothing");
    }

    // TTL expiry is unchanged by the LRU layer: an expired entry is not served.
    #[tokio::test]
    async fn expired_entry_is_not_served() {
        let cache = Cache::new(8);
        cache
            .put("a".to_string(), metadata(), Duration::from_millis(10))
            .await;
        assert!(cache.get("a").await.is_some(), "fresh before TTL");
        tokio::time::sleep(Duration::from_millis(30)).await;
        assert!(cache.get("a").await.is_none(), "expired after TTL");
    }
}
